SOME/IP input plugin

anthonypayne-GM opened the pull request:
This change introduces a new input plugin, in_someip. This plugin is used to subscribe to events and capture notifications from a SOME/IP network. It also can be used in inject SOME/IP requests and capture the response to the data pipeline. 

The input plugin uses an open source implementation of the SOME/IP stack, [vsomeip](https://github.com/COVESA/vsomeip). This library is used unchanged, but a wrapper library, libsomeip-c, exposing C-language APIs is part of the PR.

For reference to SOME/IP: [protocol specification](https://www.autosar.org/fileadmin/standards/R22-11/FO/AUTOSAR_PRS_SOMEIPProtocol.pdf).

**Testing**
Before we can approve your change; please submit the following in a comment:

- [X] Example configuration file for the change
- [X] Debug log output from testing the change
- [X] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.

- [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [N/A] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
- [N/A ] Documentation required for this feature

**Backporting**
- [N/A] Backport to latest stable release.

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

anthonypayne-GM commented:
Example config file is included in the PR, `conf/in_someip.conf`:

```
[INPUT]
    Name someip
    Tag  in.someip

    # Events to subscribe to.
    # Each event should have form:
    #    Event <service id>,<instance id>,<event id>,<event group 1>,...
    #
    # Each event must have at least one event group
    Event 4,1,32768,1
    Event 4,1,32769,2

    # RPC to send on startup
    # Each RPC entry should have form:
    #    RPC <service id>,<instance id>,<method id>,<Request Payload>
    #
    # Request payload should be base64 encoded
    RPC   4,1,1,CgAQAw==
```

anthonypayne-GM commented:
Debug log from runtime tests is attached.
[rt_test.log](https://github.com/user-attachments/files/17680279/rt_test.log)

anthonypayne-GM commented:
Output from valgrind:

```
[ OK ]
SUCCESS: All unit tests have passed.
==7910== 
==7910== HEAP SUMMARY:
==7910==     in use at exit: 5,741 bytes in 13 blocks
==7910==   total heap usage: 17,360 allocs, 17,347 frees, 5,723,214 bytes allocated
==7910== 
==7910== LEAK SUMMARY:
==7910==    definitely lost: 0 bytes in 0 blocks
==7910==    indirectly lost: 0 bytes in 0 blocks
==7910==      possibly lost: 0 bytes in 0 blocks
==7910==    still reachable: 5,741 bytes in 13 blocks
==7910==         suppressed: 0 bytes in 0 blocks
==7910== Rerun with --leak-check=full to see details of leaked memory
==7910== 
==7910== For lists of detected and suppressed errors, rerun with: -s
==7910== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

patrick-stephens commented on dockerfiles/Dockerfile:
```diff
@@ -141,6 +144,9 @@ RUN echo "deb http://deb.debian.org/debian bookworm-backports main" >> /etc/apt/
     liblzma5 \
     libyaml-0-2 \
     libcap2 \
+    libboost-system-dev \
+    libboost-thread-dev \
+    libboost-filesystem-dev \
```
This is including the dev libraries in the runtime image - we should just have the minimal set of libraries needed for running fluent bit, not those for building it.

anthonypayne-GM replied:
Hi @patrick-stephens, I tried to take these out, leaving only the entries above (line 57). But when I did, I was not able to build the SOME/IP input plugin and its dependent libraries from the dev container.

Likewise when I tried to replace these in this section with the "non-dev" version (`libboost-filesystem1.74.0`) the header files are not installed in the container and the build fails as well.

Please advise if there is something I am doing wrong here.

patrick-stephens replied:
Yeah I think you need to look at why, the dev libraries should only be in the build container which is fine. At runtime we need the minimal set of dependencies required to run with, otherwise we end up opening a big bag of security worms with extra dependencies.

Header files are not needed at runtime, they should not be in the runtime image.

So I think you need to ensure the `builder` stage has the dev libraries to compile everything but then the `dep-extractor` stage only includes the runtime libraries.

patrick-stephens commented on dockerfiles/Dockerfile:
```diff
@@ -221,6 +227,9 @@ RUN echo "deb http://deb.debian.org/debian bookworm-backports main" >> /etc/apt/
     libatomic1 \
     libgcrypt20 \
     libyaml-0-2 \
+    libboost-system-dev \
```
As above, do we really need the development libraries here or can we just use the runtime ones?

patrick-stephens replied:
This section is unrelated to building so changing it here should have no impact on compilation.

patrick-stephens replied:
Yes but this is not compiling anything, this is the debug image so again should just have the minimum set of packages per the `deb-extractor` stage.

patrick-stephens commented on dockerfiles/Dockerfile:
```diff
@@ -55,6 +55,9 @@ RUN echo "deb http://deb.debian.org/debian bookworm-backports main" >> /etc/apt/
     flex \
     bison \
     libyaml-dev \
+    libboost-system-dev \
+    libboost-thread-dev \
+    libboost-filesystem-dev \
```
We're including these development libraries but not actually enabling the plugin itself - is that right?

patrick-stephens replied:
Ah, we globally enabled it which may be an entirely separate issue.

anthonypayne-GM replied:
So, I removed the global enabling of the plugin. 

I kept these boost libraries here (they are required to build the vsomeip code) to allow enabling the SOME/IP input plugin (i.e. `-DFLB_IN_SOMEIP=On`)

patrick-stephens commented on lib/libsomeip-c/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+project(someipc CXX)
+
+include(FetchContent)
+FetchContent_Declare(
+    vsomeip3
+    GIT_REPOSITORY https://github.com/COVESA/vsomeip
+    GIT_TAG        0b83e24d16e1611958194e9b727136522f46556b # 3.5.1
+)
+FetchContent_MakeAvailable(vsomeip3)
```
Dependencies are generally vendored into the `/lib` directory and not fetched remotely.

patrick-stephens commented on CMakeLists.txt:
```diff
@@ -227,6 +227,7 @@ if(FLB_ALL)
   set(FLB_IN_NGINX_STATUS 1)
   set(FLB_IN_EXEC         1)
   set(FLB_IN_UNIX_SOCKET  1)
+  set(FLB_IN_SOMEIP       1)
```
I don't think you can enable this globally without considering all platforms that are supported including Windows and macOS configuration.

patrick-stephens commented on lib/libsomeip-c/example/test_service.cc:
```diff
@@ -0,0 +1,138 @@
+/*  Fluent Bit
```
This seems to be a C++ file, is that just a standard example and/or can it be C only? It seems to require a C++ compiler by looking at the example cmake linkage.

patrick-stephens reviewed (changes requested):
Thanks for the contribution, I think we need to look at those runtime dependencies though at the very least - also do we need to include them in the target builds as well under `./packaging/distros`?

I don't think we can realistically globally enable this for every supported target. It also seems to use C++ compilation and dependencies.

patrick-stephens commented:
For new plugins we must ensure it builds for all targets and CI so I've triggered that. It looks like we have a lot of failures around missing dependencies and the like in various places/platforms.

anthonypayne-GM commented:
Hi @patrick-stephens ,

In order to develop this plugin, I needed an update to the Docker image used by the devcontainer (`.devcontainer/devcontainer.json`) to include the boost dev libraries.

```
	"name": "devcontainer-fluent-bit",
	"image": "fluent/fluent-bit:latest-debug",
```

To get an update to this image, I modified `./dockerfiles/Dockerfile` as per the PR and locally installed the image with the boost dev libraries (`$ docker build -t fluent/fluent-bit:latest-debug -f dockerfiles/Dockerfile .`).

The dev-container uses the docker image to build fluent bit so that is why I need the dev versions of the boost library in the container image.

It appears from the comments though, that the Docker file I updated in the PR is not the correct one for updating the docker image used by the dev container?

Can you please provide some guidance on how I should proceed?  

Also, I am not sure about the Docker files that are under `packaging/distros`. Should those be updated with the plugin dependencies as well?

anthonypayne-GM commented:
> For new plugins we must ensure it builds for all targets and CI so I've triggered that. It looks like we have a lot of failures around missing dependencies and the like in various places/platforms.

This plugin should only be built for linux environments (no MacOS or Windows). Since this plugin requires boost to be installed on Ubuntu do I need to update the `.github/workflows/unit-tests.yaml` file to add in the apt install of the required boost libraries?

For Centos 7, the available boost from yum repo is to old to support this plugin (vsomeip requires boost 1.66 or newer; yum repo has boost 1.53 I think).

ammolitor commented:
@patrick-stephens I'd like to help get this over the line. Is there something you need to start for CI builds so we can get feedback and address any issues? The only changes at this point is to get the PR up-to-date with current the `master` branch. 

Thanks!

patrick-stephens commented:
> @patrick-stephens I'd like to help get this over the line. Is there something you need to start for CI builds so we can get feedback and address any issues? The only changes at this point is to get the PR up-to-date with current the `master` branch.
> 
> Thanks!

Apologies, missed some updates but I think all my previous comments still stand.
We cannot be adding -dev libraries to runtime images and we need to minimise any dependencies added to only those required.

The `packaging/distros` container definitions need updating to allow this plugin to be built for relevant targets or disabled appropriately if not possible.

ammolitor commented:
Thanks @patrick-stephens we'll work through these, and update.

patrick-stephens commented:
> * [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.

Take a look at `[N/A] Run [local packaging test](https://github.com/fluent/fluent-bit/pull/packaging/local-build-all.sh) showing all targets (including any new ones) build`.

You can build individual targets via `./packaging/build.sh -d <distro>`, e.g. `./packaging/build.sh -d centos/7`. I can label the PR as well to build all targets as it would do for a release.

ammolitor commented:
> > * [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
> 
> Take a look at [`./packaging/local-build-all.sh`](https://github.com/fluent/fluent-bit/blob/master/packaging/local-build-all.sh).
> 
> You can build individual targets via `./packaging/build.sh -d <distro>`, e.g. `./packaging/build.sh -d centos/7`. I can label the PR as well to build all targets as it would do for a release.

Thanks for pointing out something that should have been obvious 🤦‍♂️

ammolitor commented:
okay I see a test failure, but the error message is `Error: The operation was canceled.` After digging a bit it i _think_ that `flb-rt-core_chunk_trace` might have taken too long. Is that possible? and is it related to our changes?

patrick-stephens commented:
> okay I see a test failure, but the error message is `Error: The operation was canceled.` After digging a bit it i _think_ that `flb-rt-core_chunk_trace` might have taken too long. Is that possible? and is it related to our changes?

It looks like it hit the 60 minute timeout: https://github.com/fluent/fluent-bit/actions/runs/12817347279/job/35783077195?pr=9570

Not sure why, sometimes weirdness happens with CI so retriggering.

ammolitor commented:
It looks like the current test failure is a Kubernetes setup issue. Am I reading that correctly? is there anything else blocking this?

patrick-stephens commented:
Not sure as no K8S is actually used for unit tests so I'll get @pwhelan to have a look at the specific K8S event test in detail. It's on macOS which has been flaky before so likely just a test flake.

ammolitor commented:
Hi @patrick-stephens and @pwhelan just pinging to see if there is anything I can do to help move this forward. Thanks!

patrick-stephens commented:
> Hi @patrick-stephens and @pwhelan just pinging to see if there is anything I can do to help move this forward. Thanks!

Please just be patient, there are a lot of things to review so it will be in the queue.

pwhelan commented:
> Not sure as no K8S is actually used for unit tests so I'll get @pwhelan to have a look at the specific K8S event test in detail. It's on macOS which has been flaky before so likely just a test flake.

As far as I can tell it is a macOS flakey test. The tests on macOS can usually be more adversely affected by noisy neighbors at time. I am going to re-run it to see if that allows it to pass. I will also make a note to take a look at it when I can to fix it.

IMHO, it should not block this PR, @patrick-stephens and others.

pwhelan commented:
I re-ran the job and it passed. I will make a note, as I said earlier, to see why the test flaked out. I suspect it has a similar problem to other tests I have fixed before.

ammolitor commented:
bringing PR branch up to date with latest `master` branch.

pwhelan commented:
The fix for the flakey test has been submitted: https://github.com/fluent/fluent-bit/pull/9877. 

It has affected other PRs so it is a known quantity for the team.

ammolitor commented:
Hi @patrick-stephens just making sure you don't forget about this :blush:. Please let me know if there is anything I can do (besides be patient 😆) to move this forward.

patrick-stephens commented:
Need to look at the container push failures - my suspicion is down to running on a fork so no secrets to use but it should be using ephemeral ones anyway. The actual builds seem to be ok though.

patrick-stephens commented:
Can we add a docs PR with YAML config as well in here as an example? We need YAML config for all new stuff now.

patrick-stephens commented:
As per my previous comment let's add YAML config as that is the new standard now.

We need to actually have this in the builds too - at the moment it is default disabled so is never built. Can we add it to the Dockerfile and any relevant distros under `packaging` so we can show it building?

anthonypayne-GM commented:
@patrick-stephens,

I updated the [docs PR](https://github.com/fluent/fluent-bit-docs/pull/1569) to include both legacy and yaml configuration example. 

Do I also need to add a yaml conf file under `fluent-bit/conf` as well? I didn't see any yaml conf files in this repo.

patrick-stephens commented:
> @patrick-stephens,
> 
> I updated the [docs PR](https://github.com/fluent/fluent-bit-docs/pull/1569) to include both legacy and yaml configuration example.
> 
> Do I also need to add a yaml conf file under `fluent-bit/conf` as well? I didn't see any yaml conf files in this repo.

If you're adding a legacy config then yes a YAML config too.

patrick-stephens commented on dockerfiles/Dockerfile:
```diff
@@ -63,6 +63,32 @@ RUN echo "deb http://deb.debian.org/debian bookworm-backports main" >> /etc/apt/
+    rm -rf /tmp/*
+
+# fetch and compile vsomeip version 3.5.1
+RUN curl -Ls https://github.com/COVESA/vsomeip/archive/refs/tags/3.5.1.tar.gz -o vsomeip.tar.gz && \
+    echo "220ff5f105ac5f7eb33071986a12505526415217b28d4dd75164b38087dc9bcc  vsomeip.tar.gz" > vsomeip-sha.txt && \
+    sha256sum --quiet -c vsomeip-sha.txt && \
+    mkdir -p /opt/vsomeip && \
+    tar --strip-components 1 --no-same-owner --directory /opt/vsomeip -xzf vsomeip.tar.gz && \
+    cd /opt/vsomeip && \
+    cmake -Bbuild -H. && cd build && make -j "$(getconf _NPROCESSORS_ONLN)" install && \
+    rm -rf /tmp/*
```
We should be vendoring this code in I think under `/lib` - relying on some external dependency like this is not great and not the normal way it is done by Fluent Bit.

patrick-stephens commented on dockerfiles/Dockerfile:
```diff
@@ -198,6 +225,12 @@ COPY --from=deb-extractor /dpkg /
 # Copy certificates
 COPY --from=builder /etc/ssl/certs /etc/ssl/certs
 
+# Copy vsomeip
+COPY --from=builder-vsomeip /usr/local/lib/libvsomeip* /usr/local/lib/
+COPY --from=builder-vsomeip /usr/local/lib/cmake/vsomeip3/* /usr/local/lib/cmake/vsomeip3/
+COPY --from=builder-vsomeip /usr/local/lib/pkgconfig/vsomeip3.pc /usr/local/lib/pkgconfig/
+COPY --from=builder-vsomeip ./usr/local/include/vsomeip/. /usr/local/include/vsomeip/
+
```
If we vendor the code in then it removes this brittle copying as the library should be packaged ideally statically or as a direct dependency, plus with all its dependencies identified. 

A `COPY` like this is brittle as it does not capture any transitive dependencies and potentially breaks at runtime (not build time or test time) for production images if anything changes in the future.
